packages/graph/hash_graph/lib/graph/src/store/postgres/query/database.rs

```diff
@@ -0,0 +1,272 @@
+use std::{borrow::Cow, fmt, fmt::Write};
+
+use serde::Serialize;
+
+use crate::store::postgres::query::Render;
+
+/// A table available in the database.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Table {
+    TypeIds,
+}
+
+impl Table {
+    /// Returns the [`Column`] used for joining another `Table` on this `Table`.
+    pub const fn source_join_column(self) -> Column<'static> {
+        let column = match self {
+            Self::TypeIds => "version_id",
+        };
+        Column {
+            table: self,
+            access: ColumnAccess::Table { column },
+        }
+    }
+
+    /// Returns the [`Column`] used for joining this `Table` on another `Table`.
+    pub const fn target_join_column(self) -> Column<'static> {
+        let column = match self {
+            Self::TypeIds => "version_id",
+        };
+        Column {
+            table: self,
+            access: ColumnAccess::Table { column },
+        }
+    }
+}
+
+impl Render for Table {
+    fn render(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_char('"')?;
+        self.serialize(&mut *fmt)?;
+        fmt.write_char('"')
+    }
+}
+
+/// A table available in a compiled query.
+///
+/// When joining tables, each table requires a unique name. A `TableRef` may either be a plain
+/// [`Table`], or a [`Table`] with additional information attached to uniquely identify the table.
+///
+/// # Examples
+///
+/// When specifying multiple conditions or deeply nested queries containing the same [`Table`],
+/// [`Alias`] uniquely identifies the condition and the depth of the query.
+///
+/// ## Multiple Conditions
+///
+/// When searching for a [`PropertyType`], which should contain two different [`DataType`]s
+/// specified by it's `$id`, `Table::DataTypes` has to be joined twice, but with different
+/// conditions.
+///
+/// ## Deeply nested query chains
+///
+/// It's possible to have queries, which requires the same table multiple times in a chain. When
+/// searching for an [`PropertyType`], which contains a [`PropertyType`], which contains another
+/// [`PropertyType`], the `Table::PropertyTypePropertyTypeReferences` has to be joined twice
+/// within the same condition.
+///
+/// [`Alias`]: Self::Alias
+/// [`DataType`]: type_system::DataType
+/// [`PropertyType`]: type_system::PropertyType
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum TableRef {
+    /// A table inside of a compiled query, which was not renamed.
+    Table { table: Table },
+    /// A aliased table inside of a compiled query.
+    Alias {
+        table: Table,
+        condition_index: usize,
```

Comment:
I described these fields in the examples above and will also move these into another struct in one of my next PRs (as those parameters should live together all the time). I'll move over the examples to that struct then so it will be more clear.

Reply:
The PR is up:
- https://github.com/hashintel/hash/pull/1208